src/realm/sync/subscriptions.cpp

```diff
@@ -305,6 +306,64 @@ SubscriptionSet SubscriptionSet::make_mutable_copy() const
     return new_set_obj;
 }
 
+util::Future<void> SubscriptionSet::get_state_change_notification(State notify_when) const
+{
+    // If we've already reached the desired state, or if the subscription is in an error state,
+    // we can return a ready future immediately.
+    if (state() >= notify_when) {
+        return util::Future<void>::make_ready();
+    }
+    else if (state() == State::Error) {
+        return util::Future<void>::make_ready(Status{ErrorCodes::RuntimeError, error_str()});
+    }
+
+    std::lock_guard<std::mutex> lk(m_mgr->m_pending_notifications_mutex);
+
+    // If we've already been superceded by another version getting completed, then we should skip registering
+    // a notification because it may never fire.
+    if (m_mgr->m_min_outstanding_version > version()) {
+        return util::Future<void>::make_ready();
+    }
+
+    // Otherwise, make a promise/future pair and add it to the list of pending notifications.
+    auto [promise, future] = util::make_promise_future<void>();
+    m_mgr->m_pending_notifications.emplace_back(version(), std::move(promise), notify_when);
+    return std::move(future);
+}
+
+void SubscriptionSet::process_notifications()
```

Comment:
Since the API allows deleting a SubscriptionSet, how do you want to handle outstanding notifications on a subscription that gets erased? Right now it looks like they will never complete.

Reply:
The API doesn't allow deleting an arbitrary SubscriptionSet - but marking one as complete removes all the prior ones. This is what the `m_min_outstanding_version` variable is for. When a SubscriptionSet gets marked as Complete, it will go through and resolve all the Promises for SubscriptionSets with versions less than itself and bump `m_min_outstanding_version` to itself so that if you do request a notification for a `SubscriptionSet` that has now been deleted it will resolve immediately. That happens on line 324 of this patch.